stimpacks/aws/login.go

```diff
@@ -0,0 +1,161 @@
+		return err
+	}
+	log.Debug("Account: ", account, " Role: ", role)
+
+	envSource := a.stim.GetConfigBool("env-source")
+	stsLogin := a.stim.GetConfigBool("aws-web")
+	if stsLogin && a.stim.IsAutomated() {
+		a.stim.Fatal(errors.New("IsAutomated is detected: web login can not be used."))
+	}
+
+	secret, err := a.vault.AWScredentials(account, role, stsLogin)
+	if err != nil {
+		return err
+	}
+
+	if stsLogin {
+		loginURL, err := createAWSLoginURL(secret)
+		if err != nil {
+			return err
+		}
+
+		err = open.Run(loginURL)
+		if err != nil {
+			return err
+		}
+	} else {
+		if envSource { // Used for setting AWS credentials in the current environment
+			fmt.Println("export AWS_ACCESS_KEY_ID=" + secret.Data["access_key"].(string))
+			fmt.Println("export AWS_SECRET_ACCESS_KEY=" + secret.Data["secret_key"].(string))
+		} else {
+			fmt.Println("AWS_ACCESS_KEY_ID=" + secret.Data["access_key"].(string))
+			fmt.Println("AWS_SECRET_ACCESS_KEY=" + secret.Data["secret_key"].(string))
+		}
+	}
+
+	return nil
+}
+
+// createAWSLoginURL returns a federation AWS URL used for wev console login
+// This uses AWS Security Token Service (AWS STS) AssumeRole
+// More info at: https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_enable-console-custom-url.html
+// Thanks to Lachlan Donald for the following code: https://github.com/99designs/aws-vault
+func createAWSLoginURL(secret *api.Secret) (string, error) {
+	region := ""
+	path := ""
+	loginURLPrefix, destination := createRegionalURL(region, path)
+
+	req, err := http.NewRequest("GET", loginURLPrefix, nil)
+	if err != nil {
+		return "", err
+	}
+
+	// Note: This AWS API doesn't validate given info
+	jsonBytes, err := json.Marshal(map[string]string{
+		"sessionId":    secret.Data["access_key"].(string),
+		"sessionKey":   secret.Data["secret_key"].(string),
+		"sessionToken": secret.Data["security_token"].(string),
+	})
+	if err != nil {
+		return "", err
+	}
+
+	q := req.URL.Query()
+	q.Add("Action", "getSigninToken")
+	q.Add("Session", string(jsonBytes))
+
+	req.URL.RawQuery = q.Encode()
+
+	// Note: You can still get a token if you have the wrong credentials
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return "", errors.New("Failed to create federated token: " + err.Error())
+	}
+
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return "", errors.New("Call to getSigninToken failed with " + resp.Status)
+	}
+
+	var respParsed map[string]string
+
+	if err = json.Unmarshal([]byte(body), &respParsed); err != nil {
+		return "", errors.New("Failed to parse response from getSigninToken: " + err.Error())
+	}
+
+	signinToken, ok := respParsed["SigninToken"]
+	if !ok {
+		return "", errors.New("Expected a response with SigninToken")
+	}
+
+	loginURL := fmt.Sprintf(
+		"%s?Action=login&Issuer=%s&Destination=%s&SigninToken=%s",
+		loginURLPrefix,
+		url.QueryEscape(stimURL),
```

Comment:
Not sure why we would want to link to github at all, is this required for some reason? If there was a reason I would think it should be a configurable link at the very least.

Reply:
Ya, this is not needed. I was looking at the python sts tool that does the same thing and thought it looked nice. Maybe it would be a good idea to have a global config that all of "these" variables could be moved to. I will look around for a clean what of doing that with a later change. I will add a TODO in the code to track that.